_posts/2018-07-20-exploring-state-management-in-react-container-components.md

```diff
@@ -0,0 +1,122 @@
+
+<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">You Might Not Need Redux <a href="https://t.co/3zBPrbhFeL">https://t.co/3zBPrbhFeL</a></p>&mdash; Dan Abramov (@dan_abramov) <a href="https://twitter.com/dan_abramov/status/777983404914671616?ref_src=twsrc%5Etfw">September 19, 2016</a></blockquote>
+<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
+
+In this series, we'll explore a few different patterns you can introduce to your code base before reaching for a 3rd party solution!
+
+## Container/Presenter Components
+
+This pattern separates what might be a single component into two (or three or four), a Container component(s) to maintain state and a Presenter component to render visual markup. 
+
+```jsx
+const Cart = props => (
+  <ul>
+    <h2>Cart</h2>
+    <input
+      type="text"
+      onChange={props.onNewItemChange}
+      value={props.newItem}
+    />
+    <button onClick={props.addItem}>Add item</button>
+
+    {props.items.map((item, idx) => (
+      <li key={idx}>
+        <button onClick={props.removeItem(item)}>Remove</button>
+        {item}, &ensp;
+        {props.discounts[idx]}% off!
+      </li>
+    ))}
+  </ul>
+);
+
+class CartContainer extends React.Component {
+  state = {
+    newItem: "",
+    items: [],
+    discounts: []
+  };
+
+  componentDidUpdate(prevProps, prevState) {
+    if (prevState.items.length !== this.state.items.length) {
+      const itemsQuery = this.state.items.join(",");
+
+      ajax(`https://discountdb.com/?items=${itemsQuery}`)
+        .then(discounts => this.setState({ discounts }));
+    }
+  }
+
+  onNewItemChange = ({target:{value: newItem}}) => this.setState({newItem});
+
+  addItem = () => {
+    this.setState(prevState => ({
+      items: [...prevState.items, prevState.newItem],
+      newItem: ""
+    }));
+  };
+
+  removeItem = item => () => {
+    this.setState(prevState => ({
+      items: prevState.items.filter(
+        (i, idx, prevItems) => idx !== prevItems.indexOf(item)
+      )
+    }));
+  };
+
+  render() {
+    return (
+      <div>
+        <Cart
+          newItem={this.state.newItem}
+          onNewItemChange={this.onNewItemChange}
+          items={this.state.items}
+          discounts={this.state.discounts}
+          addItem={this.addItem}
+          removeItem={this.removeItem}
+        />
+      </div>
+    );
+  }
+}
+```
+[![Edit wo7y9voowk](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/wo7y9voowk)
+&nbsp;
+
+Here you can see that we have a Container component that handles controlling component state with the `addItem`, `removeItem`, and `onNewItemChange` callbacks, and fetching a list of discounts from an external REST api. This enables us to write `Cart` as a Pure Functional component. 
```

Comment:
💡 'This enables us to write `Cart`, our Presenter component, as a Pure Functional component.'